File: 3. Longest Substring Without Repeating Characters.java
public class Solution {
    public int lengthOfLongestSubstring(String s) {
        if(s == null || s.length() == 0)    return 0;
        
        Map<Character, Integer> map = new HashMap<>();
        int res = 0;
        for(int i = 0, j = 0; i < s.length(); i ++) {
            char chr = s.charAt(i);
            if(map.containsKey(chr)) {
                j = Math.max(j, map.get(chr) + 1);
            }
            map.put(chr, i);
            res = Math.max(res, i - j + 1);
        }
        
        return res;
    }
}

File: 3. Longest Substring Without Repeating Characters.cpp
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        unordered_map<char, int> map;
        int res = 0;
        
        for(int i = 0, j = 0; i < s.size(); i ++) {            
            if(map.find(s[i]) != map.end()) {
                j = max(j, (map.find(s[i])->second));
            }
            
            res = max(res, i - j + 1);
            map[s[i]] = i + 1;
        }
        
        return res;
    }
};
